ichnaea/service/locate.py

```diff
@@ -122,496 +100,536 @@ def map_data(data):
-
-    if len(clusters) == 0:
-        stats_client.incr('%s.wifi.found_no_cluster' % api_name)
-        return None
-
-    clusters.sort(lambda a, b: cmp(len(b), len(a)))
-    cluster = clusters[0]
-    sample = cluster[:min(len(cluster), MAX_WIFIS_IN_CLUSTER)]
-    length = len(sample)
-    avg_lat = sum([n.lat for n in sample]) / length
-    avg_lon = sum([n.lon for n in sample]) / length
-    return {
-        'lat': avg_lat,
-        'lon': avg_lon,
-        'accuracy': estimate_accuracy(avg_lat, avg_lon,
-                                      sample, WIFI_MIN_ACCURACY),
-    }
+                for j in range(len(clusters)):
+                    if merged_one:
+                        break
+                    if i == j:
+                        continue
+                    a = clusters[i]
+                    b = clusters[j]
+                    if cluster_distance(a, b) <= threshold:
+                        clusters.pop(j)
+                        a.extend(b)
+                        merged_one = True
+
+        return [[items[i] for i in c] for c in clusters]
+
+    def filter_bssids_by_similarity(self, bs):
+        # Cluster BSSIDs by "similarity" (hamming or arithmetic distance);
+        # return one BSSID from each cluster. The distance threshold is
+        # hard-wired to 2, meaning that two BSSIDs are clustered together
+        # if they are within a numeric difference of 2 of one another or
+        # a hamming distance of 2.
+
+        DISTANCE_THRESHOLD = 2
+
+        def bytes_of_hex_string(hs):
+            return [int(hs[i:i + 2], 16) for i in range(0, len(hs), 2)]
+
+        def hamming_distance(a, b):
+            h = 0
+            v = a ^ b
+            while v:
+                h += 1
+                v &= v - 1
+            return h
+
+        def hamming_or_arithmetic_byte_difference(a, b):
+            return min(abs(a - b), hamming_distance(a, b))
+
+        def bssid_difference(a, b):
+            abytes = bytes_of_hex_string(a)
+            bbytes = bytes_of_hex_string(b)
+            return sum(hamming_or_arithmetic_byte_difference(a, b) for
+                       (a, b) in zip(abytes, bbytes))
+
+        clusters = self.cluster_elements(
+            bs, bssid_difference, DISTANCE_THRESHOLD)
+        return [c[0] for c in clusters]
+
+    def get_clean_wifi_keys(self, data):
+        wifis = []
+
+        # Pre-process wifi data
+        for wifi in data.get('wifi', ()):
+            wifi = normalized_wifi_dict(wifi)
+            if wifi:
+                wifis.append(wifi)
+
+        # Estimate signal strength at -100 dBm if none is provided,
+        # which is worse than the 99th percentile of wifi dBms we
+        # see in practice (-98).
+
+        wifi_signals = dict([(w['key'], w['signal'] or -100) for w in wifis])
+        wifi_keys = set(wifi_signals.keys())
+
+        return wifis, wifi_signals, wifi_keys
+
+    def query_database(self, wifi_keys):
+        queried_wifis = []
+        try:
```

Comment:
I'd put the check here:
```
if len(wifi_keys) < MIN_WIFIS_IN_QUERY:
    return []

queried_wifis = []
try:
```
At this point it won't screw up the stats, and prevents us from doing an accidental query for all wifi keys. As a plus it avoids doing a db query when we only have a single wifi and won't return a result based on it, no matter whether we find the one or not.